src/input/mpegts/linuxdvb/linuxdvb_ddci.c

```diff
@@ -835,21 +935,74 @@ linuxdvb_ddci_open ( linuxdvb_ddci_t *lddci )
 }
 
 void
-linuxdvb_ddci_put ( linuxdvb_ddci_t *lddci, const uint8_t *tsb, int len )
+linuxdvb_ddci_put
+  ( linuxdvb_ddci_t *lddci, service_t *t, elementary_stream_t *st,
+    const uint8_t *tsb, int len )
 {
-  linuxdvb_ddci_wr_thread_buffer_put(&lddci->lddci_wr_thread, tsb, len );
+  int_fast16_t pid = st ? st->es_pid : -1;
+
+  if (linuxdvb_ddci_cam_pid_filter(pid))
```

Comment:
Can we replace this 'filtering' by this code:

<pre>
if (lddci->prev_tsb == tsb) return;
lddci->prev_tsb  = tsb;
linuxdvb_ddci_wr_thread_buffer_put(&lddci->lddci_wr_thread, tsb, len );
</pre>

Reply:
The code you suggested will ignore the same data pointer.
I remember now an eMail from you telling me, that some PIDs (0, 1, ...) are sent in a loop to all services of a mux.
So yes, this code is required to not duplicate them when sending to the CAM.

But we can't exchange the whole code with that, because I want to send all the PIDs immediately to the service **and** only those **also** to the CAM, which the CAM really needs. And this are only a few of the special PIDs.
When those PIDs come out of the CAM we simply **don't** forward them to the service again, because this has been done already in linuxdvb_ddci_put.
The reason for that is, that for MTD we need to change the PIDs in the PAT (and all the other packets) and it is much easier to remove the modified PAT when it comes out of the CAM, than changing it back. Moreover the PAT is already perfect for the service when linuxdvb_ddci_put is called. And there will be different PATs for different tuners. So the new artificial PAT needs to be discarded anyway and then there would be no PAT in the stream to the service. 

And I plan another feature with sending an empty EIT to the CAM to get rid of the PIN question. For that the EIT would be not correct when it comes out of the CAM to be sent to a service. So it needs to be send in linuxdvb_ddci_put to the service. Moreover, there will be different EITs in case of MTD and when this comes out of the CAM, we don't know for which service the EIT is (no pID mapping for special PIDs possible). So the only working solution is to discard all this special PIDs at the output and forward them to the right service at the input.
Or does you see another solution?

Reply:
> But we can't exchange the whole code with that, because I want to send all the PIDs immediately to the service and only those also to the CAM, which the CAM really needs. And this are only a few of the special PIDs.

I won't do that. You reorder the input data and I'd prefer to retain the original order for the 'pass' output (where the MPEG-TS data are just passed to the client). For MTD, the situation is more complicated, but you can probably use the SI section index to distinguish which original PAT should be used on output (after DDCI).